cf/regrid/regrid.py

```diff
@@ -0,0 +1,1895 @@
+                    n, m = b.shape[0:2]
+                    tmp = np.empty((n + 1, m + 1), dtype=b.dtype)
+                    tmp[:n, :m] = b[:, :, 0]
+                    tmp[:n, m] = b[:, -1, 1]
+                    tmp[n, :m] = b[-1, :, 3]
+                    tmp[n, m] = b[-1, -1, 2]
+                    bounds[dim] = tmp
+
+    # Define the ESMF.Grid stagger locations
+    if bounds:
+        if n_axes == 3:
+            staggerlocs = [
+                ESMF.StaggerLoc.CENTER_VCENTER,
+                ESMF.StaggerLoc.CORNER_VFACE,
+            ]
+        else:
+            staggerlocs = [ESMF.StaggerLoc.CORNER, ESMF.StaggerLoc.CENTER]
+    else:
+        if n_axes == 3:
+            staggerlocs = [ESMF.StaggerLoc.CENTER_VCENTER]
+        else:
+            staggerlocs = [ESMF.StaggerLoc.CENTER]
+
+    # Create an empty ESMF.Grid
+    esmf_grid = ESMF.Grid(
+        max_index=np.array(shape, dtype="int32"),
+        coord_sys=coord_sys,
+        num_peri_dims=num_peri_dims,
+        periodic_dim=periodic_dim,
+        staggerloc=staggerlocs,
+    )
+
+    # Populate the ESMF.Grid centres
+    for dim, c in enumerate(coords):
+        if n_axes == 3:
+            grid_centre = esmf_grid.get_coords(
+                dim, staggerloc=ESMF.StaggerLoc.CENTER_VCENTER
+            )
+        else:
+            grid_centre = esmf_grid.get_coords(
+                dim, staggerloc=ESMF.StaggerLoc.CENTER
+            )
+
+        grid_centre[...] = c
+
+    # Populate the ESMF.Grid corners
+    if bounds:
+        if n_axes == 3:
+            staggerloc = ESMF.StaggerLoc.CORNER_VFACE
+        else:
+            staggerloc = ESMF.StaggerLoc.CORNER
+
+        for dim, b in enumerate(bounds):
+            grid_corner = esmf_grid.get_coords(dim, staggerloc=staggerloc)
+            grid_corner[...] = b
+
+    #    # Add an ESMF.Grid mask
+    #    if mask is not None:
+    #        m = None
+    #        mask = np.asanyarray(mask)
+    #        if np.ma.isMA(mask):
+    #            if np.ma.is_masked(mask):
+    #                m = mask.mask
+    #        elif mask.dtype == bool and mask.any():
+    #            m = mask
+    #
+    #        if m is not None:
+    #            grid_mask = esmf_grid.add_item(ESMF.GridItem.MASK)
+    #            # Note: ESMF mask has False for masked elements.
+    #            grid_mask[...] = np.invert(m).astype("int32")
+
+    # Add an ESMF.Grid mask
+    if mask is not None:
+        if mask.dtype != bool:
+            raise ValueError(
+                "'mask' must be None or a Boolean array. Got: "
+                f"dtype={mask.dtype}"
+            )
+
+        if not mask.any():
+            mask = None
+
+        if mask is not None:
+            grid_mask = esmf_grid.add_item(ESMF.GridItem.MASK)
+            if len(grid.coords) == 2 and mask.ndim == 1:
+                # ESMF grid has a dummy size 1 dimension, so we need to
+                # include this int he mas as well.
```

Comment:
```suggestion
                # include this in the mask as well.
```

Great approach by creating these dummy dimensions, BTW.